File: iot/amqp/header/impl/AMQPOpen.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.HeaderCode as HeaderCode
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVList as TLVList

class amqpOpen():
    def __init__(self,code,doff,type,channel,containerId,hostname,maxFrameSize,channelMax,idleTimeout,outgoingLocales,incomingLocales,offeredCapabilities,desiredCapabilities,properties):
        if code is None:
            self.code = HeaderCode.headerCode.getValueByKey('OPEN')
        else:
            self.code = code

        if doff is not None:
            self.doff = doff
        else:
            self.doff = 2

        if type is not None:
            self.type = type
        else:
            self.type = 0

        if channel is not None:
            self.channel = channel
        else:
            self.channel = 0

        self.containerId = containerId
        self.hostname = hostname
        self.maxFrameSize = maxFrameSize
        self.channelMax = channelMax
        self.idleTimeout = idleTimeout
        self.outgoingLocales = outgoingLocales
        self.incomingLocales = incomingLocales
        self.offeredCapabilities = offeredCapabilities
        self.desiredCapabilities = desiredCapabilities
        self.properties = properties

    def setDoff(self,doff):
        self.doff = doff

    def getDoff(self):
        return self.doff

    def getType(self):
        return self.type

    def setType(self, type):
        self.type = type

    def getChannel(self):
        return self.channel

    def setChannel(self, channel):
        self.channel = channel

    def getCode(self):
        return self.code

    def toArgumentsList(self):
        list = TLVList.tlvList(None,None)
        wrapper = AMQPWrapper.amqpWrapper()
        if self.containerId is None:
            print("Detach header's container id can't be null")
        list.addElement(0, wrapper.wrap(self.containerId))
        if self.hostname is not None:
            list.addElement(1, wrapper.wrap(self.hostname))
        if self.maxFrameSize is not None:
            list.addElement(2, wrapper.wrap(self.maxFrameSize))
        if self.channelMax is not None:
            list.addElement(3, wrapper.wrap(self.channelMax))
        if self.idleTimeout is not None:
            list.addElement(4, wrapper.wrap(self.idleTimeout))
        if self.outgoingLocales is not None and len(self.outgoingLocales) > 0:
            list.addElement(5, wrapper.wrapArray(self.outgoingLocales))
        if self.incomingLocales is not None and len(self.incomingLocales) > 0:
            list.addElement(6, wrapper.wrapArray(self.incomingLocales))
        if self.offeredCapabilities is not None and len(self.offeredCapabilities) > 0:
            list.addElement(7, wrapper.wrapArray(self.offeredCapabilities))
        if self.desiredCapabilities is not None and len(self.desiredCapabilities) > 0:
            list.addElement(8, wrapper.wrapArray(self.desiredCapabilities))
        if self.properties is not None and len(self.properties) > 0:
            list.addElement(9, wrapper.wrapMap(self.properties))

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(AMQPType.amqpType.getValueByKey('SMALL_ULONG'), self.code.value))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        size = len(list.getList())
        if size  == 0:
            print("Received malformed Open header: container id can't be null")
        if size > 10:
            print("Received malformed Open header. Invalid number of arguments: " + str(size))

        element = list.getList()[0]

        if element is None:
            print("Received malformed Open header: container id can't be null")
        self.containerId = unwrapper.unwrapString(element)

        if size > 1:
            element = list.getList()[1]
            if element is not None and not element.isNull():
                self.hostname = unwrapper.unwrapString(element)
        if size > 2:
            element = list.getList()[2]
            if element is not None and not element.isNull():
                self.maxFrameSize = unwrapper.unwrapUInt(element) 
        if size > 3:
            element = list.getList()[3]
            if element is not None and not element.isNull():
                self.channelMax = unwrapper.unwrapUShort(element)
        if size > 4:
            element = list.getList()[4]
            if element is not None and not element.isNull():
                self.idleTimeout = unwrapper.unwrapUInt(element)
        if size > 5:
            element = list.getList()[5]
            if element is not None and not element.isNull():
                self.outgoingLocales = unwrapper.unwrapArray(element)
        if size > 6:
            element = list.getList()[6]
            if element is not None and not element.isNull():
                self.incomingLocales = unwrapper.unwrapArray(element)
        if size > 7:
            element = list.getList()[7]
            if element is not None and not element.isNull():
                self.offeredCapabilities = unwrapper.unwrapArray(element)
        if size > 8:
            element = list.getList()[8]
            if element is not None and not element.isNull():
                self.desiredCapabilities = unwrapper.unwrapArray(element)
        if size > 9:
            element = list.getList()[9]
            if element is not None and not element.isNull():
                self.properties = unwrapper.unwrapMap(element)

    def toString(self):
        return "AMQPOpen [containerId=" + str(self.containerId) + ", hostname=" + str(self.hostname) + ", maxFrameSize=" + str(self.maxFrameSize) + ", channelMax=" + str(self.channelMax) + ", idleTimeout=" + str(self.idleTimeout) + ", outgoingLocales=" + str(self.outgoingLocales) + ", incomingLocales=" + str(self.incomingLocales) + ", offeredCapabilities=" + str(self.offeredCapabilities) + ", desiredCapabilities=" + str(self.desiredCapabilities) + ", properties=" + str(self.properties) + ", code=" + str(self.code) + ", doff=" + str(self.doff) + ", type=" + str(self.type) + ", channel=" + str(self.channel) + "]"

    def setContainerId(self, containerId):
        self.containerId = containerId

    def getContainerId(self):
        return self.containerId

    def setHostname(self, hostname):
        self.hostname = hostname

    def getHostname(self):
        return self.hostname

    def setMaxFrameSize(self, maxFrameSize):
        self.maxFrameSize = maxFrameSize

    def getMaxFrameSize(self):
        return self.maxFrameSize

    def setChannelMax(self, channelMax):
        self.channelMax = channelMax

    def getChannelMax(self):
        return self.channelMax

    def setIdleTimeout(self, idleTimeout):
        self.idleTimeout = idleTimeout

    def getIdleTimeout(self):
        return self.idleTimeout

    def setOutgoingLocales(self, outgoingLocales):
        self.outgoingLocales = outgoingLocales

    def getOutgoingLocales(self):
        return self.outgoingLocales

    def setIncomingLocales(self, incomingLocales):
        self.incomingLocales = incomingLocales

    def getIncomingLocales(self):
        return self.incomingLocales

    def setOfferedCapabilities(self, offeredCapabilities):
        self.offeredCapabilities = offeredCapabilities

    def getOfferedCapabilities(self):
        return self.offeredCapabilities

    def setDesiredCapabilities(self, desiredCapabilities):
        self.desiredCapabilities = desiredCapabilities

    def getDesiredCapabilities(self):
        return self.desiredCapabilities

    def setProperties(self, properties):
        self.properties = properties

    def getProperties(self):
        return self.properties

File: IoTbrokerClient/src/iot-protocols/amqp/classes/headeramqp/amqpopen.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpopen.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvfixed.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvnull.h"
#include "iot-protocols/amqp/classes/amqpsimpletype.h"

AMQPOpen::AMQPOpen() : AMQPHeader(new AMQPHeaderCode(AMQP_OPEN_HEADER_CODE))
{
    this->maxFrameSize = NULL;
    this->channelMax = NULL;
    this->idleTimeout = NULL;
    this->outgoingLocales = list<JsonVariant *>();
    this->incomingLocales = list<JsonVariant *>();
    this->offeredCapabilities = list<JsonVariant *>();
    this->desiredCapabilities = list<JsonVariant *>();
    this->properties = map<JsonVariant *, JsonVariant *>();
}

int AMQPOpen::getLength()
{
    int length = 8;
    AMQPTLVList *arguments = this->arguments();
    length += arguments->getLength();
    return length;
}

int AMQPOpen::getType()
{
    return AMQP_OPEN_HEADER_CODE;
}

AMQPTLVList *AMQPOpen::arguments()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (this->containerId == NULL) {
        printf("AMQPOpen::arguments::containerId");
        return new AMQPTLVList();
    }
    list->addElementWithIndex(0, AMQPWrapper::wrapString(this->containerId));

    if (this->hostname != NULL) {
        list->addElementWithIndex(1, AMQPWrapper::wrapString(this->hostname));
    }
    if (this->maxFrameSize != NULL) {
        list->addElementWithIndex(2, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->maxFrameSize)));
    }
    if (this->channelMax != NULL) {
        list->addElementWithIndex(3, AMQPWrapper::wrapUShort(AMQPSimpleType::variantToUShort(this->channelMax)));
    }
    if (this->idleTimeout != NULL) {
        list->addElementWithIndex(4, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->idleTimeout)));
    }
    if (this->outgoingLocales.count() != 0) {
        list->addElementWithIndex(5, AMQPWrapper::wrapArray(this->outgoingLocales));
    }
    if (this->incomingLocales.count() != 0) {
        list->addElementWithIndex(6, AMQPWrapper::wrapArray(this->incomingLocales));
    }
    if (this->offeredCapabilities.count() != 0) {
        list->addElementWithIndex(7, AMQPWrapper::wrapArray(this->offeredCapabilities));
    }
    if (this->desiredCapabilities.count() != 0) {
        list->addElementWithIndex(8, AMQPWrapper::wrapArray(this->desiredCapabilities));
    }
    if (this->properties.count() != 0) {
        list->addElementWithIndex(9, AMQPWrapper::wrapMap(this->properties));
    }

    unsigned char * data = malloc(sizeof(char));
    data[0] = this->code->getValue();
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);
    return list;
}

void AMQPOpen::fillArguments(AMQPTLVList *list)
{
    int size = list->getList().count();

    if (size == 0) {
        printf("AMQPOpen::fillArguments::size == 0");
        return;
    }
    if (size > 10) {
        printf("AMQPOpen::fillArguments::wrong_type");
        return;
    }

    TLVAMQP *element = list->getList()[0];
    if (element->isNull()) {
        printf("AMQPOpen::fillArguments::containerId");
    }
    this->containerId = AMQPUnwrapper::unwrapString(element);

    if (size > 1) {
        element = list->getList()[1];
        if (!element->isNull()) {
            this->hostname = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (size > 2) {
        element = list->getList()[2];
        if (!element->isNull()) {
            this->maxFrameSize = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
        }
    }
    if (size > 3) {
        element = list->getList()[3];
        if (!element->isNull()) {
            this->channelMax = AMQPSimpleType::UShortToVariant(AMQPUnwrapper::unwrapUShort(element));
        }
    }
    if (size > 4) {
        element = list->getList()[4];
        if (!element->isNull()) {
            this->idleTimeout = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
        }
    }
    if (size > 5) {
        element = list->getList()[5];
        if (!element->isNull()) {
            this->outgoingLocales = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (size > 6) {
        element = list->getList()[6];
        if (!element->isNull()) {
            this->incomingLocales = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (size > 7) {
        element = list->getList()[7];
        if (!element->isNull()) {
            this->offeredCapabilities = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (size > 8) {
        element = list->getList()[8];
        if (!element->isNull()) {
            this->desiredCapabilities = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (size > 9) {
        element = list->getList()[9];
        if (!element->isNull()) {
            this->properties = AMQPUnwrapper::unwrapMap(element);
        }
    }
}


void AMQPOpen::addOutgoingLocale(list<String> array)
{
    std::list<String>::iterator it;
    for (it = array.begin(); it != array.end(); ++it){
        JsonVariant *variant;
        variant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(it)));
        this->outgoingLocales.push_back(variant);
    }
}

void AMQPOpen::addIncomingLocale(list<String> array)
{
    std::list<String>::iterator it;
    for (it = array.begin(); it != array.end(); ++it){
        JsonVariant *variant;
        variant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(it)));
        this->incomingLocales.push_back(variant);
    }
}

void AMQPOpen::addOfferedCapability(list<String> array)
{
    std::list<String>::iterator it;
    for (it = array.begin(); it != array.end(); ++it){
        JsonVariant *variant;
        variant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(it)));
        this->offeredCapabilities.push_back(variant);
    }
}

void AMQPOpen::addDesiredCapability(list<String> array)
{
    std::list<String>::iterator it;
    for (it = array.begin(); it != array.end(); ++it){
        JsonVariant *variant;
        variant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(it)));
        this->desiredCapabilities.push_back(variant);
    }
}

void AMQPOpen::addProperty(String key, variant *value)
{
    JsonVariant *variantKey;
    variantKey.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(key)));
    this->properties.insert(variantKey, value);
}

String AMQPOpen::getContainerId() const
{
    return containerId;
}

void AMQPOpen::setContainerId(const String &value)
{
    containerId = value;
}

String AMQPOpen::getHostname() const
{
    return hostname;
}

void AMQPOpen::setHostname(const String &value)
{
    hostname = value;
}

JsonVariant *AMQPOpen::getMaxFrameSize() const
{
    return maxFrameSize;
}

void AMQPOpen::setMaxFrameSize(JsonVariant *value)
{
    maxFrameSize = value;
}

JsonVariant *AMQPOpen::getChannelMax() const
{
    return channelMax;
}

void AMQPOpen::setChannelMax(JsonVariant *value)
{
    channelMax = value;
}

JsonVariant *AMQPOpen::getIdleTimeout() const
{
    return idleTimeout;
}

void AMQPOpen::setIdleTimeout(JsonVariant *value)
{
    idleTimeout = value;
}

list<JsonVariant *> AMQPOpen::getOutgoingLocales() const
{
    return outgoingLocales;
}

void AMQPOpen::setOutgoingLocales(const list<JsonVariant *> &value)
{
    outgoingLocales = value;
}

list<JsonVariant *> AMQPOpen::getIncomingLocales() const
{
    return incomingLocales;
}

void AMQPOpen::setIncomingLocales(const list<JsonVariant *> &value)
{
    incomingLocales = value;
}

list<JsonVariant *> AMQPOpen::getOfferedCapabilities() const
{
    return offeredCapabilities;
}

void AMQPOpen::setOfferedCapabilities(const list<JsonVariant *> &value)
{
    offeredCapabilities = value;
}

list<JsonVariant *> AMQPOpen::getDesiredCapabilities() const
{
    return desiredCapabilities;
}

void AMQPOpen::setDesiredCapabilities(const list<JsonVariant *> &value)
{
    desiredCapabilities = value;
}

map<JsonVariant *, JsonVariant *> AMQPOpen::getProperties() const
{
    return properties;
}

void AMQPOpen::setProperties(const map<JsonVariant *, JsonVariant *> &value)
{
    properties = value;
}
